### PR TITLE
Use stateful set for pods with storage

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

For units requiring storage, use a k8s StatefulSet instead of a Deployment Controller to manage the pods.
This PR required a regeneration of the mocks using go generate so ignore those file when reviewing.

The basis is to use a deployment controller for pods without storage and a statefulset for those with. The storage setup logic is essentially the same, but is tweaked a bit because the pv claim creation is not done  manually for stateful sets.

## QA steps

Deploy a test k8s charm with storage. Ensure a pv and pvc is properly set up and the specified mount point used.

### CODE REVIEW

File: caas/kubernetes/provider/k8s.go [hunk 1]
```diff
@@ -178,9 +181,31 @@ func (k *kubernetesClient) EnsureOperator(appName, agentPath string, config *caa
...
+				Spec: *pvcSpec,
+			})
+			if err != nil {
+				return errors.Trace(err)
+			}
+			logger.Debugf("created new pvc: %+v", pvc)
+		}
+		storageVol = &core.Volume{Name: volName}
+		storageVol.PersistentVolumeClaim = &core.PersistentVolumeClaimVolumeSource{
+			ClaimName: pvc.Name,
```
Comment (ycliuhw) [inline caas/kubernetes/provider/k8s.go]: Do we need to handle `NotFound` here?

File: caas/kubernetes/provider/k8s.go [hunk 2]
```diff
@@ -468,17 +481,19 @@ func (k *kubernetesClient) EnsureService(
...
-		return errors.Annotatef(err, "configuring storage for %s", appName)
-	}
-
 	// Add a deployment controller configured to create the specified number of units/pods.
 	numPods := int32(numUnits)
-	if err := k.configureDeployment(appName, unitSpec, params.PodSpec.Containers, &numPods); err != nil {
-		return errors.Annotate(err, "creating or updating deployment controller")
+	if len(params.Filesystems) > 0 {
+		if err := k.configureStatefulSet(appName, unitSpec, params.PodSpec.Containers, &numPods, params.Filesystems); err != nil {
+			return errors.Annotate(err, "creating or updating stateful set")
```
Comment (ycliuhw) [inline caas/kubernetes/provider/k8s.go]: is it `the StatefulSet` here?

**Comment by ycliuhw** [comment]
It's much better to use `StatefulSet` to ensure our stateful operator app. 
Nice!

File: caas/kubernetes/provider/k8s.go (hunk 2 above)
Comment (wallyworld) [reply caas/kubernetes/provider/k8s.go]: yeah, so if the unit requires storage, we use a stateful set
if there's no storage required, we continue to use a deployment controller as it's lighter weight

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**